analyzers/src/SonarAnalyzer.Common/Trackers/ArgumentTracker.cs

```diff
@@ -0,0 +1,83 @@
+    protected abstract bool InvocationFitsMemberKind(SyntaxNode argumentNode, InvokedMemberKind memberKind);
+    protected abstract bool InvokedMemberFits(SemanticModel model, SyntaxNode argumentNode, InvokedMemberKind memberKind, Func<string, bool> invokedMemberNameConstraint);
+    protected abstract SyntaxNode InvokedExpression(SyntaxNode argumentNode);
+
+    public Condition MatchArgument(ArgumentDescriptor descriptor) =>
+        context =>
+        {
+            if (context.Node is { } argumentNode
+                && SyntacticChecks(context.SemanticModel, descriptor, argumentNode)
+                && InvokedExpression(argumentNode) is { } invoked
+                && MethodSymbol(context.SemanticModel, invoked) is { } methodSymbol
+                && Language.MethodParameterLookup(invoked, methodSymbol).TryGetSymbol(argumentNode, out var parameter)
+                && ParameterFits(parameter, descriptor.ParameterConstraint, descriptor.InvokedMemberConstraint))
+            {
+                context.Parameter = parameter;
+                return true;
+            }
+            return false;
+        };
+
+    private IMethodSymbol MethodSymbol(SemanticModel model, SyntaxNode invoked) =>
+        model.GetSymbolInfo(invoked).Symbol switch
+        {
+            IMethodSymbol x => x,
+            IPropertySymbol propertySymbol => Language.Syntax.IsWrittenTo(invoked, model, CancellationToken.None)
+                ? propertySymbol.SetMethod
+                : propertySymbol.GetMethod,
+            _ => null,
+        };
+
+    private bool SyntacticChecks(SemanticModel model, ArgumentDescriptor descriptor, SyntaxNode argumentNode) =>
+        InvocationFitsMemberKind(argumentNode, descriptor.MemberKind)
+        && (descriptor.RefKind is not { } expectedRefKind || (ArgumentRefKind(argumentNode) is not { } actualRefKind) || actualRefKind == expectedRefKind)
+        && (descriptor.ArgumentListConstraint == null
+            || (ArgumentList(argumentNode) is { } argList && descriptor.ArgumentListConstraint(argList, Position(argumentNode))))
+        && (descriptor.InvokedMemberNameConstraint == null
+            || InvokedMemberFits(model, argumentNode, descriptor.MemberKind, x => descriptor.InvokedMemberNameConstraint(x, Language.NameComparison)));
+
+    private static bool ParameterFits(IParameterSymbol parameter, Func<IParameterSymbol, bool> parameterConstraint, Func<IMethodSymbol, bool> invokedMemberConstraint)
+    {
+        if ((parameter.ContainingSymbol is IMethodSymbol method)
+            && method.Parameters.IndexOf(parameter) is >= 0 and int position)
+        {
+            do
+            {
+                if (invokedMemberConstraint?.Invoke(method) is null or true && parameterConstraint?.Invoke(method.Parameters[position]) is null or true)
+                {
+                    return true;
+                }
+            }
+            while ((method = method.OverriddenMethod) != null);
```

Comment:
Educational: I don't understand this iteration:
- where is `method` updated inside the do loop?
- I might be wrong again but I cannot see a test case for this, for example I can rewrite this method like:
```cs
private static bool ParameterFits(IParameterSymbol parameter, Func<IParameterSymbol, bool> parameterConstraint, 
Func<IMethodSymbol, bool> invokedMemberConstraint) =>
        parameter.ContainingSymbol is IMethodSymbol method
        && method.Parameters.IndexOf(parameter) is >= 0 and int position
        && invokedMemberConstraint?.Invoke(method) is null or true
        && parameterConstraint?.Invoke(method.Parameters[position]) is null or true;
```

And it's working fine against UTs

Reply:
Method_Inheritance_BaseClasses_Overrides, Method_Inheritance_BaseClasses_Generics and the VB counterparts are failing if you do that.

> where is method updated inside the do loop?

Here: `((method = method.OverriddenMethod) != null)`

`method` gets assigned and the result of the assignment is compared against null. An assignment is an expression statement. That means it can be used as an expression or a statement, just like invocations:
```cs
SomeAction(); // invocation used as statement
var x = SomeCalculation(); // invocation used as expression

x = 5; // assignment used as statement
y = (x = 5); // assignment x = 5 used as expression. 
```

See also https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/assignment-operator